File: memOShield/memoshield/notifier.py
import os
import logging
import requests

TELEGRAM_TOKEN = os.getenv('TELEGRAM_BOT_TOKEN')
TELEGRAM_CHAT = os.getenv('TELEGRAM_CHAT_ID')
SLACK_WEBHOOK = os.getenv('SLACK_WEBHOOK')


class Notifier:
    def __init__(self):
        self.token = TELEGRAM_TOKEN
        self.chat_id = TELEGRAM_CHAT
        self.slack = SLACK_WEBHOOK

    def send_telegram(self, text: str) -> bool:
        if not (self.token and self.chat_id):
            logging.debug('Telegram creds not set; skipping notification')
            return False
        url = f'https://api.telegram.org/bot{self.token}/sendMessage'
        payload = {'chat_id': self.chat_id, 'text': text}
        try:
            r = requests.post(url, json=payload, timeout=5)
            return r.ok
        except Exception as e:
            logging.warning('Failed to send telegram: %s', e)
            return False

    def send_slack(self, text: str) -> bool:
        if not self.slack:
            logging.debug('Slack webhook not set; skipping')
            return False
        try:
            r = requests.post(self.slack, json={'text': text}, timeout=5)
            return r.ok
        except Exception as e:
            logging.warning('Failed to send slack message: %s', e)
            return False

    def notify_critical(self, event: dict) -> None:
        """Send critical alert to configured channels. Non-blocking best-effort."""
        text = f"[CRITICAL] {event.get('type')} from {event.get('ip')} reason={event.get('reason')}"
        try:
            # best-effort, ignore failures
            self.send_telegram(text)
            self.send_slack(text)
        except Exception:
            logging.exception('Notifier failed')

File: memOShield/internal/notifier/notifier.go
package notifier

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/MemoS778/memOShield/internal/config"
)

// Notifier sends alerts to Telegram and Slack.
type Notifier struct {
	token   string
	chatID  string
	slack   string
	client  *http.Client
}

// New creates a Notifier.
func New() *Notifier {
	return &Notifier{
		token:  config.TelegramBotToken,
		chatID: config.TelegramChatID,
		slack:  config.SlackWebhook,
		client: &http.Client{Timeout: 5 * time.Second},
	}
}

// SendTelegram sends a message via Telegram Bot API.
func (n *Notifier) SendTelegram(text string) bool {
	if n.token == "" || n.chatID == "" {
		log.Println("Telegram creds not set; skipping notification")
		return false
	}
	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", n.token)
	payload, _ := json.Marshal(map[string]string{
		"chat_id": n.chatID,
		"text":    text,
	})
	resp, err := n.client.Post(url, "application/json", bytes.NewReader(payload))
	if err != nil {
		log.Printf("Failed to send telegram: %v", err)
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == http.StatusOK
}

// SendSlack sends a message via Slack webhook.
func (n *Notifier) SendSlack(text string) bool {
	if n.slack == "" {
		log.Println("Slack webhook not set; skipping")
		return false
	}
	payload, _ := json.Marshal(map[string]string{"text": text})
	resp, err := n.client.Post(n.slack, "application/json", bytes.NewReader(payload))
	if err != nil {
		log.Printf("Failed to send slack message: %v", err)
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == http.StatusOK
}

// NotifyCritical sends alerts to all configured channels (best-effort).
func (n *Notifier) NotifyCritical(eventType, ip, reason string) {
	text := fmt.Sprintf("[CRITICAL] %s from %s reason=%s", eventType, ip, reason)
	go n.SendTelegram(text)
	go n.SendSlack(text)
}
